### PR TITLE
executor: refine `explain analyze`

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

this PR address the contiune comments in #7827 

the mainly change are:

- protect runtimeStats with mutex, `explain analyze` is infrequently operation but executor maybe uder concurrent execution.
- forbidden collect stats of the IndexLookupExecutor that owned by  table worker, it seems no other executor like this one - -
- refactor explainBuilder/explainExec, push render data & analyze logic down to explainExec
- rename some variable/struct

### What is changed and how it works?

- RT

### Check List 

Tests 

 - Old test

Code changes

 - impl refactor

Side effects

 - no

Related changes

 - Need to cherry-pick to 2.1

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7888)

### CODE REVIEW

**Comment by lysu** [comment]
/run-all-tests

&&

PTAL @zz-jason

File: util/execdetails/execdetails.go [hunk 1]
```diff
@@ -69,43 +72,35 @@ type RuntimeStat struct {
 }
 
 // NewRuntimeStats creates new executor collector.
-func NewRuntimeStats() RuntimeStats {
-	return RuntimeStats(make(map[string]*RuntimeStat))
+func NewRuntimeStats() *RuntimeStatsColl {
```
Comment (zz-jason) [inline util/execdetails/execdetails.go]: how about: s/NewRuntimeStats/NewRuntimeStatsColl/

File: util/execdetails/execdetails.go [hunk 2]
```diff
@@ -69,43 +72,35 @@ type RuntimeStat struct {
...
+	return &RuntimeStatsColl{stats: make(map[string]*RuntimeStats)}
 }
 
 // GetRuntimeStat gets execStat for a executor.
-func (e RuntimeStats) GetRuntimeStat(planID string) *RuntimeStat {
+func (e *RuntimeStatsColl) GetRuntimeStat(planID string) *RuntimeStats {
```
Comment (zz-jason) [inline util/execdetails/execdetails.go]: how about s/GetRuntimeStat/Get/

File: util/execdetails/execdetails.go [hunk 3]
```diff
@@ -69,43 +72,35 @@ type RuntimeStat struct {
...
 	if e == nil {
```
Comment (zz-jason) [inline util/execdetails/execdetails.go]: this check can be removed?

File: executor/distsql.go [hunk 4]
```diff
@@ -489,7 +488,9 @@ func (e *IndexLookUpExecutor) buildTableReader(ctx context.Context, handles []in
 		feedback:        statistics.NewQueryFeedback(0, nil, 0, false),
 		corColInFilter:  e.corColInTblSide,
 		plans:           e.tblPlans,
-	}, handles)
+	}
+	tableReaderExec.statsEnable = false
```
Comment (zz-jason) [inline executor/distsql.go]: should be set according to the statement type, i.e.:
```go
ctx.GetSessionVars().StmtCtx.RuntimeStatsColl != nil
```

File: executor/executor.go [hunk 5]
```diff
@@ -78,7 +78,8 @@ type baseExecutor struct {
 	maxChunkSize  int
 	children      []Executor
 	retFieldTypes []*types.FieldType
-	runtimeStat   *execdetails.RuntimeStat
+	statsEnable   bool
+	_runtimeStats *execdetails.RuntimeStats
```
Comment (zz-jason) [inline executor/executor.go]: I think the old way is better. And to eliminate misunderstanding that `e.runtimeStats` will always be set, in function `newBaseExecutor()` we can:
```go
if ctx.GetSessionVars().StmtCtx.RuntimeStatsColl != nil {
	e.runtimeStats = e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRuntimeStat(e.id)
}
```

Comment (lysu) [reply executor/executor.go]: Done~

**Comment by zz-jason** [comment]
@XuHuaiyu @lamxTyler PTAL

File: util/execdetails/execdetails.go [hunk 6]
```diff
@@ -68,44 +71,33 @@ type RuntimeStat struct {
...
 	atomic.AddInt64(&e.rows, int64(rowNum))
 }
 
-func (e *RuntimeStat) String() string {
+func (e *RuntimeStats) String() string {
 	if e == nil {
 		return ""
 	}
-	return fmt.Sprintf("time:%f, loops:%d, rows:%d", time.Duration(e.consume).Seconds()*1e3, e.loop, e.rows)
+	return fmt.Sprintf("time:%fms, loops:%d, rows:%d", time.Duration(e.consume).Seconds()*1e3, e.loop, e.rows)
```
Comment (lamxTyler) [inline util/execdetails/execdetails.go]: May be we can use `%v` to let the system libaray handle the time unit.

Comment (lysu) [reply util/execdetails/execdetails.go]: IMHO, fixed `ms` is better.

Comment (lamxTyler) [reply util/execdetails/execdetails.go]: If it takes too long, we need to mentally parse it to something like `min` or `hour`?

Comment (lysu) [reply util/execdetails/execdetails.go]: it's not normal has min or hour executue time  and more 000000 is ok for me too :rofl:

how do u think about ~? @zz-jason  @XuHuaiyu

Comment (lysu) [reply util/execdetails/execdetails.go]: DONE~~~PTAL @lamxTyler  :laughing:

**Pull request merged**